aesara/compile/builders.py

```diff
@@ -43,38 +43,38 @@ def infer_shape(outs, inputs, input_shapes):
             assert len(inp_shp) == inp.type.ndim
 
     shape_feature = ShapeFeature()
-    shape_feature.on_attach(FunctionGraph([], []))
+    fgraph = FunctionGraph([], [])
+    shape_feature.on_attach(fgraph)
 
     # Initialize shape_of with the input shapes
     for inp, inp_shp in zip(inputs, input_shapes):
-        shape_feature.set_shape(inp, inp_shp)
+        shape_feature.set_shape(fgraph, inp, inp_shp)
 
     def local_traverse(out):
         """
         Go back in the graph, from out, adding computable shapes to shape_of.
 
         """
-        if out in shape_feature.shape_of:
+        if out in fgraph.shape_of:
             # Its shape is already known
```

Comment:
Is `shape_of` now always present in a FunctionGraph, or only when the `ShapeFeature` is around?  Or was it already?

Reply:
Under these changes, `shape_of` is now added to a `FunctionGraph` whenever `ShapeFeature` is attached.  Currently, this `shape_of` `dict` resides in the `ShapeFeature` instance attached to `FunctionGraph`s.

Reply:
Could this pose any issues with pickling the FunctionGraph?

I see you added a bunch of pickling tests but didn't notice one for this.

Reply:
Basically, the changes in this PR put the additional features added by `Feature`s onto the `FunctionGraph`s themselves.  `Feature`s then become more like dynamic mixin classes.  Really, this whole `Feature` idea could be recast as `FunctionGraph` subclassing, since all of the functionality they can add is more directly implementable by extending `FunctionGraph` via a subclass.

Reply:
> Could this pose any issues with pickling the FunctionGraph?
> 
> I see you added a bunch of pickling tests but didn't notice one for this.

I'm not sure how `pickle`ing could affect this situation, since both the `FunctionGraph` and its `ShapeFeature` are only created at runtime within this `infer_shape` function.